connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java

```diff
@@ -0,0 +1,189 @@
+import org.apache.parquet.schema.MessageType;
+import org.teiid.file.VirtualFile;
+import org.teiid.file.VirtualFileConnection;
+import org.teiid.language.LanguageObject;
+import org.teiid.metadata.RuntimeMetadata;
+import org.teiid.translator.DataNotAvailableException;
+import org.teiid.translator.Execution;
+import org.teiid.translator.ExecutionContext;
+import org.teiid.translator.TranslatorException;
+
+
+public class BaseParquetExecution implements Execution {
+    @SuppressWarnings("unused")
+    protected ExecutionContext executionContext;
+    @SuppressWarnings("unused")
+    protected RuntimeMetadata metadata;
+    protected VirtualFileConnection connection;
+    protected boolean immutable;
+
+    // Execution state
+    protected ParquetQueryVisitor visitor = new ParquetQueryVisitor();
+    private VirtualFile[] parquetFiles;
+    private AtomicInteger fileCount = new AtomicInteger();
+    private ParquetFileReader reader;
+    private MessageType schema;
+    private RecordReader<Group> rowIterator;
+    private MessageColumnIO columnIO;
+    private long pageRowCount;
+
+    public BaseParquetExecution(ExecutionContext executionContext,
+                                RuntimeMetadata metadata, VirtualFileConnection connection, boolean immutable) {
+        this.executionContext = executionContext;
+        this.metadata = metadata;
+        this.connection = connection;
+        this.immutable = immutable;
+    }
+
+    public void visit(LanguageObject command) throws TranslatorException {
+        this.visitor.visitNode(command);
+
+        if (!visitor.exceptions.isEmpty()) {
+            throw visitor.exceptions.get(0);
+        }
+    }
+
+    @Override
+    public void execute() throws TranslatorException {
+        this.parquetFiles = VirtualFileConnection.Util.getFiles(this.visitor.getParquetPath(), this.connection, true);
+        VirtualFile nextParquetFile = getNextParquetFile();
+        if (nextParquetFile != null) {
+            this.rowIterator = readParquetFile(nextParquetFile);
+        }
+    }
+
+    private RecordReader<Group> readParquetFile(VirtualFile parquetFile) throws TranslatorException {
+        try (InputStream parquetFileStream = parquetFile.openInputStream(!immutable)) {
+           return readParquetFile(parquetFile, parquetFileStream);
+        } catch (IOException e) {
+            throw new TranslatorException(e);
+        }
+    }
+
+    private RecordReader<Group> readParquetFile(VirtualFile parquetFile, InputStream parquetFileStream) throws TranslatorException {
+       try {
+           File localFile = createTempFile(parquetFileStream);
+           Path path = new Path(localFile.toURI());
+           Configuration config = new Configuration();
+           reader = ParquetFileReader.open(HadoopInputFile.fromPath(path, config));
+           schema = reader.getFooter().getFileMetaData().getSchema();
+           columnIO = new ColumnIOFactory().getColumnIO(schema);
```

Comment:
Once you have the expected columns, you can use that to read with a filtered schema.  That is you will construct a new MesssageType here that contains only fields that have been requested.  See the ColumnIOFactory.getColumnIO that take a requested schema argument.  I would guess that you would pass the filtered schema to the GroupRecordConverter as well.

Reply:
I'll check this out! Thanks!